[Feature] New module `geospatial-client` for cross-plugin IP enrichment

andy-k-improving opened the pull request:
### Description
This is an implementation ticket which follow the discussion of https://github.com/opensearch-project/geospatial/issues/698, to create a new Gradle module for other OpenSearch Plugin to import, in order to make an IP enrichment call to GeoSpatial plugin to resolve GeoLocation data with existing data source. 

Hight-level code changes:
 - A new Gradle module `geospatial-client` with appreciate Gradle setting files to publish it as a standalone Jar. 
 - A core class `ipEnrichmentActionClient` which serve as the interface, this is the only way for other Plugin to interact with GeoSpatial's IP enrichment functionality
 - A collection of classes to support the IP enrichment remote call access `IpEnrichmentAction`, `IpEnrichmentRequest` and `IpEnrichmentResponse`
 - `IpEnrichementTransportAction` class which contain the logic for the IP address lookup and register it during the GeoSpatial's bootstrap time


Non GeoSpatial plugin can now interact with the new functionality like below:
```
IpEnrichmentActionClient ipClient = new IpEnrichmentActionClient(client.getNodeClient());
        try {
          Map<String, Object> geoLocationData = ipClient.getGeoLocationData("50.68.18.229");
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
```


### Related Issues
Resolves https://github.com/opensearch-project/geospatial/issues/698


### Check List
- [x] New functionality includes testing.
- [ ] New functionality has been documented.
- [ ] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md).
- [x] Commits are signed per the DCO using `--signoff`.
- [ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/geospatial/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

andy-k-improving commented:
@heemin32 here is the implementation for https://github.com/opensearch-project/geospatial/issues/698 as we discussed earlier of the week, would you mind to have a look and advise?
Any comment will be appreciated.

navneet1v commented on client/build.gradle:
```diff
@@ -0,0 +1,58 @@
+allprojects {
+    group = opensearch_group
+    version = "${opensearch_build}"
+    targetCompatibility = JavaVersion.VERSION_11
+    sourceCompatibility = JavaVersion.VERSION_11
```
is this compatibility correct? I thought whole opensearch moved to java 17.

andy-k-improving replied:
Actually I think that make more sense to enforce the same from `opensearch.build` instead, in this way the Java version requirement of this module will always identical to OpenSearch main repo.

andy-k-improving replied:
I replaced above with `apply plugin: 'opensearch.build'` instead.

navneet1v commented on client/src/test/java/org/opensearch/geospatial/action/IpEnrichmentActionClientTest.java:
```diff
@@ -0,0 +1,57 @@
+import static org.mockito.ArgumentMatchers.eq;
+import static org.mockito.Mockito.when;
+
+@ExtendWith(MockitoExtension.class)
+class IpEnrichmentActionClientTest {
```
For all the tests lets use the same test framework used in plugin rather than using standard Mockito ensure that test framework is same across different modules of the plugin.

andy-k-improving replied:
I have updated the build.gradle to read the Mockito version from the `opensearch.build` plugin, in this case we will have the same version of Mockito as the main OpenSearch repo.
```
testImplementation "org.mockito:mockito-core:${versions.mockito}"
```
Or you mean I should ditch Mockito all together?

heemin32 replied:
It seems we don't extend from `OpenSearchTestCase` for client module? https://github.com/opensearch-project/ml-commons/blob/main/client/src/test/java/org/opensearch/ml/client/MachineLearningNodeClientTest.java

navneet1v replied:
The problem with not doing this would be that we first we will not able to test the code with random data and another problem is since we will not have seed whenever we want to reproduce a failure. Not sure why MLCommons didn't do this. This is an acceptable practice in Opensearch Plugins. Plus you will miss out on all the different mocking and testing framework added by Opensearch.

andy-k-improving replied:
In my point, this depends on the amount of logic to test. 
I agree we should have it to remain the consistency, but for now if we aim to have it as a thin client which has no actual logic inside except the argument validation.
This may be a overkill.

navneet1v commented on client/src/main/java/org/opensearch/geospatial/action/IpEnrichmentActionClient.java:
```diff
@@ -0,0 +1,56 @@
+@AllArgsConstructor
+public class IpEnrichmentActionClient {
+
+    NodeClient nodeClient;
```
any reason for NodeClient being package private?

andy-k-improving replied:
Nope, updated with `final private` :)

navneet1v commented on client/build.gradle:
```diff
@@ -0,0 +1,58 @@
+    maven { url "https://plugins.gradle.org/m2/" }
+}
+
+compileJava {
```
can we add spotlessJava task or it is already imported from root module?

andy-k-improving replied:
Done, by adding it on the build.gralde explicitl.

navneet1v commented on client/src/main/java/org/opensearch/geospatial/action/IpEnrichmentAction.java:
```diff
@@ -0,0 +1,24 @@
+ * An ActionType registered on OpenSearch registry, for inter-cluster transportAction call,
+ * to resolve GeoLocation for IP String.
+ */
+public class IpEnrichmentAction extends ActionType<ActionResponse> {
```
not sure why this action is in client? and its transport action is in root module. Please correct me if I am wrong here. 
One client implementation in Opensearch project I have seen is this: https://github.com/opensearch-project/ml-commons/tree/main/client .

andy-k-improving replied:
This is needed because during the plugin runtime, other plugin need to submit the `ActionType` as an identifier for one of the argument on `nodeClient.execute(ActionType, xxx, xx)`.
I have also checked on the above ML-common integration, indeed they also expose the respective action also `MLPredictionTaskAction`, but they decided on a separate module call `opensearch-ml-common` instead. 

I think both approach work, either we can keep everything in a jar and only expose `geoSpatial-client`, or we can follow the ML plugin approach, which expose TWO jars `ml-common` and `ml-client`, but essential contains the same classes.

navneet1v reviewed (commented):
Looked very high level on the code. I like the idea of having a client. But I want to get some consistencies in the code. and I am hoping client to very lightweight.

andy-k-improving commented:
@navneet1v I have fixed most of the point except the Mockito, can you have a look a review?
Thanks,

heemin32 commented on client/src/main/java/org/opensearch/geospatial/action/IpEnrichmentActionClient.java:
```diff
@@ -0,0 +1,57 @@
+            IpEnrichmentResponse enrichmentResponse = IpEnrichmentResponse.fromActionResponse(genericActionResponse);
+            return enrichmentResponse.getGeoLocationData();
+        } catch (Exception e) {
+            log.error("GeoSpatial IP Enrichment call failure, with detail: ", e);
```
I believe we should throw an exception and let the caller handle it, rather than returning `null`?

andy-k-improving replied:
Done, updating it to throw exception instead.

heemin32 commented on src/main/java/org/opensearch/geospatial/ip2geo/action/IpEnrichmentTransportAction.java:
```diff
@@ -0,0 +1,87 @@
+        }
+    }
+
+    private String getDefaultDataSourceName() {
```
Would it be better to make the data source name mandatory? Allowing the caller to omit the data source and selecting one randomly from the available options might not be ideal.

heemin32 commented on client/src/main/java/org/opensearch/geospatial/action/IpEnrichmentResponse.java:
```diff
@@ -0,0 +1,83 @@
+     * @param actionResponse An IpEnrichmentResponse in casted-up form.
+     * @return An IpEnrichmentResponse object which contain the same payload as the incoming object.
+     */
+    public static IpEnrichmentResponse fromActionResponse(ActionResponse actionResponse) {
```
How would it work if we need to modify either the request or the response? Will these changes be handled correctly, particularly during a cluster upgrade? Additionally, could there be scenarios where other plugins use a different version of the client compared to the geospatial plugin?

andy-k-improving commented:
@heemin32 I updated the code to reflect comments, can you have another look.
Also whats your thought on the cluster upgrade scenario?

heemin32 commented:
> @heemin32 I updated the code to reflect comments, can you have another look. Also whats your thought on the cluster upgrade scenario?

I think, we can use add versioning on this serialization/deserialization using OpenSearch version.

heemin32 commented:
@Xtansia Could you take a look at this?

andy-k-improving commented:
> I think, we can add versioning on this serialization/deserialization using OpenSearch vers

Just to align on this, what's the expected behaviour in the case when  serializationID doesn't match?
Should we thrown an exception, we attempt to parse it as per the serializationID? (Ex: v1.19.0 version of the plugin serializating the object from v1.18.0...etc)

andy-k-improving commented:
> > @heemin32 I updated the code to reflect comments, can you have another look. Also whats your thought on the cluster upgrade scenario?
> 
> I think, we can add versioning on this serialization/deserialization using OpenSearch version?

I have updated the code to perform version, can you have a look and advise?
Thx!

heemin32 commented:
Another general comment is if we should implement it as a rest API rather than client? What do you think? @andy-k-improving

andy-k-improving commented on client/src/main/java/org/opensearch/geospatial/action/IpEnrichmentResponse.java:
```diff
@@ -0,0 +1,92 @@
+                if (VERSION.equals(objectVersion)) {
+                    return new IpEnrichmentResponse(input);
+                } else {
+                    throw new IllegalArgumentException("Fail to serialise IpEnrichmentResponse due to version mismatch");
```
Right, let me revert change, but to clarify on the context.
If you mean the backward competiable, we are referring something similar to https://github.com/opensearch-project/OpenSearch/blob/581771053c324e0774c3309ec9806d53d99f9f35/server/src/main/java/org/opensearch/action/admin/indices/upgrade/post/ShardUpgradeResult.java#L49

Which read the `Version` object from the constructor and use it on the serialisation || de-serailisation process to identify the version mis-match and handle the backward compatibility accordingly?

heemin32 replied:
Yes. For example. https://github.com/opensearch-project/k-NN/blob/main/src/main/java/org/opensearch/knn/index/query/parser/KNNQueryBuilderParser.java#L142-L163

andy-k-improving commented:
> Another general comment is if we should implement it as a rest API rather than client? What do you think? @andy-k-improving

We should have it in the long term || as day-2, but I don't think that make sense to bundle it in one PR, as there will be more considerations I foresee (Ex: performance, web-layer security) need to be addressed || discussed ahead of the implementation. 
Let me create another issue perhaps?

heemin32 commented on client/src/main/java/org/opensearch/geospatial/action/IpEnrichmentRequest.java:
```diff
@@ -0,0 +1,103 @@
+    public IpEnrichmentRequest(StreamInput streamInput) throws IOException {
+        super(streamInput);
+        ipString = streamInput.readString();
+        datasourceName = streamInput.readOptionalString();
```
Let's make it as non optional.

andy-k-improving replied:
Done.

andy-k-improving commented:
@heemin32 Also do you foresee any problem to publish the `geospatial-client` jar? 
In local, I see the jar is produced correctly, but not sure will that be the same when publishing to AWS || Maven central.

heemin32 commented:
> Let me create another issue perhaps?

No need for that. I was considering either this or that, but not both. Thanks.

heemin32 commented:
> @heemin32 Also do you foresee any problem to publish the `geospatial-client` jar? In local, I see the jar is produced correctly, but not sure will that be the same when publishing to AWS || Maven central.

I don’t have much expertise on that topic. @ylwu-amzn, is there anything specific we should consider when publishing the client? Asking as ml-common already has done it.

andy-k-improving commented:
> No need for that. I was considering either this or that, but not both. Thanks.

If that is the case, then I rather keep this instead of REST.

heemin32 commented on src/test/java/org/opensearch/geospatial/ip2geo/action/IpEnrichmentTransportActionTests.java:
```diff
@@ -0,0 +1,69 @@
+    }
+    
+
+    /**
```
Please remove this test case.

heemin32 replied:
Or, rename it.

andy-k-improving replied:
Removed.

andy-k-improving commented:
Rebased again for the JDK 21 changes on main, waiting for CI.

andy-k-improving commented:
@navneet1v Hi, I have updated the code and tested the build command locally, it should work by now.
Would you mind to have another look and advise?
Thanks,

andy-k-improving commented:
Updated the code style should be ok for review again.

navneet1v commented on src/main/java/org/opensearch/geospatial/ip2geo/action/IpEnrichmentTransportAction.java:
```diff
@@ -0,0 +1,67 @@
+        IpEnrichmentRequest enrichmentRequest = IpEnrichmentRequest.fromActionRequest(request);
+        String ipString = enrichmentRequest.getIpString();
+        if (enrichmentRequest.getDatasourceName() == null) {
+            log.error("No data source available, IpEnrichmentTransportAction aborted.");
```
is this error due to user configuration or related to system?

andy-k-improving replied:
User configuration, but actually this will never happen, because OpenSearch runtime should run `ipEnrichmentRequest.validate( )` ahead, so when TransportAction received the request object. 
The dataSource String should be present.

andy-k-improving commented on src/main/java/org/opensearch/geospatial/ip2geo/action/IpEnrichmentTransportAction.java:
```diff
@@ -0,0 +1,67 @@
+        String ipString = enrichmentRequest.getIpString();
+        if (enrichmentRequest.getDatasourceName() == null) {
+            log.error("No data source available, IpEnrichmentTransportAction aborted.");
+            listener.onFailure(new IllegalArgumentException());
```
Same, actually the `if` clause can be removed altogether, hence there is no need for this exception.

andy-k-improving commented:
Hi @navneet1v, would you mind to have another look?
Thanks,

heemin32 commented:
Could you rebase and see if the failing CI will succeed?

andy-k-improving commented:
@heemin32 done.

heemin32 commented:
@andy-k-improving, thanks for the change. If you don't get response from @navneet1v by next Monday, let me pull in another person to review this PR.

navneet1v commented on client/src/test/java/org/opensearch/geospatial/action/IpEnrichmentRequestTests.java:
```diff
@@ -0,0 +1,71 @@
+     */
+    @Test
+    public void testValidateValidRequest() {
+        System.out.println("Test");
```
remove this.

andy-k-improving replied:
Done.

navneet1v reviewed (commented):
Small comment. Overall code looks good to me. There is system.out statement left in the code. Added a comment for that.
